File: python/mergeSort.py
def mergeSort(unsorted_arr):
    if len(unsorted_arr) <= 1:
        return

    middle = len(unsorted_arr)//2  # [2]

    # starting from 0 till middle element
    left = unsorted_arr[:middle]


    # starting from middle to end of arra
    right = unsorted_arr[middle:]

    # (recursion) this will return sorted left and right arrays
    mergeSort(left)
    mergeSort(right)

    # return the final sorted array
    merged_sorted_array(left,right, unsorted_arr)




# to merge two sorted arrays and return a single sorted array
def merged_sorted_array(sorted_arr1, sorted_arr2, unsorted_arr):

    # specifying lengths of both sorted arrays
    len_arr1 = len(sorted_arr1)
    len_arr2 = len(sorted_arr2)

    # starting from first element
    i = j = k = 0

    # iterate through each element and compare elements
    while i < len_arr1 and j < len_arr2:
        if sorted_arr1[ i ] < sorted_arr2[ j ]:
            unsorted_arr[k] = sorted_arr1[i]
            i+=1

        else:
            unsorted_arr[k] = sorted_arr2[j]
            j += 1

        k+=1

    # if first condition is true append the element directly
    while i < len_arr1:
        unsorted_arr[k] = sorted_arr1[i]
        i += 1
        k += 1

    # if second condition is true append the element directly
    while j < len_arr2:
       unsorted_arr[k]=sorted_arr2[j]
       j += 1
       k+=1


arr = [9,4,6,2,4,8,1]
mergeSort(arr)
print(arr)

File: c++/mergeSort.cpp
#include<iostream>
using namespace std;
void merge(int a[],int low,int mid,int high,int n)
{
  int i=low;
  int j=mid+1;
  int k=low;
  int b[n];
  while(i<=mid && j<=high)
  {
    if(a[i]<a[j])
      b[k++]=a[i++];
    else
      b[k++]=a[j++];
  }
  for(;i<=mid;i++)
    b[k++]=a[i];
  for(;j<=high;j++)
    b[k++]=a[j];
  for(int i=low;i<=high;i++)
    a[i]=b[i];
}
void mergeSort(int arr[],int low,int high,int n)
{
  int mid;
  if(low<high)
  {
    mid=(low+high)/2;
    mergeSort(arr,low,mid,n);
    mergeSort(arr,mid+1,high,n);
    merge(arr,low,mid,high,n);
  }
}
void display(int a[],int n)
{
  for(int i=0;i<n;i++)
    cout<<a[i]<<" ";
}
int main() {
  int n;
  cout<<"---------------------------------------------------"<<endl;
  cout << "ENTER THE NO OF ELEMENTS:";
  cin>>n;
  int a[n]={0};
  cout<<"ENTER THE ELEMENTS:"<<endl;
  for(int i=0;i<n;i++)
  {
    cout<<"ELEMENT "<<i+1<<":";
    cin>>a[i];
  }
  cout<<"---------------------------------------------------"<<endl;
  cout<<"BEFORE SORTING:";
  display(a,n);
  mergeSort(a,0,n-1,n);
  cout<<"\nAFTER SORTING:";
  display(a,n);
  cout<<"\n---------------------------------------------------"<<endl;
  return 0;
}
